synapse/lib/httpapi.py

```diff
@@ -1240,3 +1246,182 @@ async def get(self):
+    /api/ext/.*
+    '''
+
+    storm_prefix = 'init { $request = $lib.cortex.httpapi.response($_http_request_info) }'
+
+    async def get(self, path):
+        return await self._runHttpExt('get', path)
+
+    async def head(self, path):
+        return await self._runHttpExt('head', path)
+
+    async def post(self, path):
+        return await self._runHttpExt('post', path)
+
+    async def put(self, path):
+        return await self._runHttpExt('put', path)
+
+    async def delete(self, path):
+        return await self._runHttpExt('delete', path)
+
+    async def patch(self, path):
+        return await self._runHttpExt('patch', path)
+
+    async def options(self, path):
+        return await self._runHttpExt('options', path)
+
+    async def _runHttpExt(self, meth, path):
+        core = self.getCore()
+        adef, args = await core.getHttpExtApiByPath(path)
+        if adef is None:
+            self.set_status(404)
+            self.sendRestErr('NoSuchPath', f'No Extended HTTP API endpoint matches {path}')
+            return await self.finish()
+
+        requester = None
+        useriden = adef.get('owner')
+
+        if adef.get('authenticated'):
+
+            requester = await self.useriden()
+
+            if requester is None:
+                await self.reqAuthUser()
+                return
+
+            for pdef in adef.get('perms'):
+                if not await self.allowed(pdef.get('perm'), default=pdef.get('default')):
+                    return
+
+            if adef.get('runas') == 'user':
+                useriden = requester
+
+        # We flatten the request headers and parameters into a flat key/valu map.
+        # The first instance of a given key wins.
+        request_headers = {}
+        for key, valu in self.request.headers.get_all():
+            request_headers.setdefault(key.lower(), valu)
+
+        params = {}
+        for key, valus in self.request.query_arguments.items():
+            for valu in valus:
+                params.setdefault(key, valu.decode())
+                continue
+
+        iden = adef.get("iden")
+        info = {
+            'uri': self.request.uri,
+            'body': self.request.body,
+            'iden': iden,
+            'path': path,
+            'user': requester,
+            'method': self.request.method,
+            'params': params,
+            'headers': request_headers,
+            'args': args,
+            'client': self.request.remote_ip,
+        }
+
+        varz = adef.get('vars')
+        varz['_http_request_info'] = info
+
+        opts = {
+            'readonly': adef.get('readonly'),
+            'show': (
+                'http:resp:body',
+                'http:resp:code',
+                'http:resp:headers',
+            ),
+            'user': useriden,
+            'vars': varz,
+            'view': adef.get('view'),
+        }
+
+        storm = adef['methods'].get(meth)
+
+        if storm is None:
+            self.set_status(500)
```

Comment:
seems like this should happen up with the 404 check

Reply:
I could see moving it after the authenticated check? In that default auth=true case, we'd only be sharing the lack of support for a given method AFTER we know the user was allowed to even access the endpoint.

Reply:
yeah could do that. I figured we already tell them that the endpoint exists before authenticating so knowing what methods it supports wasn't much more info.